keps/0010-dynamic-plugin-configuration.md

```diff
@@ -0,0 +1,221 @@
+| title                        | status      |
+|------------------------------|-------------|
+| Dynamic Plugin Configuration | provisional |
+
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+    - [Story 4](#story-4)
+    - [Story 5](#story-5)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+    - [Notes](#notes)
+    - [Caveats](#caveats)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently the configuration for the `KongPlugin` and `KongClusterPlugin` resources
+is generally static, meaning that it is not possible to derive configuration values from
+common configuration resources defined on same kubernetes cluster. This leads to a lot
+duplication of hard coded values in numerous places making the deployment process
+difficult to manage. While it is possible to load configuration from a secret using
+the `configFrom` parameter, the entire configuration must be contained with in the
+secret which is effectively the same problem.
+
+It would be advantageous, and more in line with common Kubernetes behavior to be able
+to derive individual values of plugin configuration from data already stored on the same
+Kubernetes cluster. This proposal aims to define a change to the `KongPlugin` and `KongClusterPlugin`
+Custom Resource Definitions to facilitate this behavior.
+
+## Motivation
+
+Allow `KongPlugin` and `KongClusterPlugin` resources to read configuration values from Kubernetes,
+in addition to literal values.
+
+Continue to allow `KongPlugin` and `KongClusterPlugin` resource to accept literal input values.
+
+Reduce the amount of duplication of configuration value in the simple case. That is,
+when using standard kubernetes objects and with out assumed templating systems.
+
+### Goals
+
+* Provide ability to read individual configuration values from `ConfigMap` resources
+* Provide ability to read individual configuration values from `Secret` resources
+* Provide ability to read individual configuration values from arbitrary or generic resource types (other custom resource types)
```

Comment:
Did you have anything specific in mind here? If not, I would ask that we keep things tight by sticking with the use cases we know. If so, perhaps there's a user-story we can add that will help support this goal?

Reply:
I suppose this really give me the ability to read data from any other CRD that may store data

```yaml
    - genericMapKeyRef:
        apiVersion: a.custom.resource/v1
        kind: ResourceFoo
        name: foo-bar
        namespace: default
```

with this sort of an implementation, doing configmap + secret key look ups aren't that much different and keep the internals consistent + simple

Reply:
This would seem to be a source of significant complexity. For the moment since it would appear this may be mostly a "want" rather than a "need" do you think it would be reasonable to add this to the alternatives considered section as something possibly for a future iteration?

Reply:
of course. secrets + configmaps are certainly the need.

I bring it up as I've implement this in [another controller](https://github.com/razee-io/razeedeploy-core/blob/86981ef2ccb61526d9bcb3848fa39f89c928a0b6/lib/FetchEnvs.js#L50-L64)  and it works out rather cleanly.